test/kotlin/integration/CancelStreamTest.kt

```diff
@@ -0,0 +1,155 @@
+              stat_prefix: remote_hcm
+              route_config:
+                name: remote_route
+                virtual_hosts:
+                - name: remote_service
+                  domains: ["*"]
+                  routes:
+                  - match: { prefix: "/" }
+                    direct_response: { status: 200 }
+              http_filters:
+              - name: envoy.router
+                typed_config:
+                  "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+      - name: base_api_listener
+        address:
+          socket_address: { protocol: TCP, address: 0.0.0.0, port_value: 10000 }
+        api_listener:
+          api_listener:
+            "@type": $hcmType
+            stat_prefix: api_hcm
+            route_config:
+              name: api_router
+              virtual_hosts:
+              - name: api
+                domains: ["*"]
+                routes:
+                - match: { prefix: "/" }
+                  route: { cluster: fake_remote }
+            http_filters:
+            - name: envoy.filters.http.platform_bridge
+              typed_config:
+                "@type": $pbfType
+                platform_filter_name: $filterName
+            - name: envoy.router
+              typed_config:
+                "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+      clusters:
+      - name: fake_remote
+        connect_timeout: 0.25s
+        type: STATIC
+        lb_policy: ROUND_ROBIN
+        load_assignment:
+          cluster_name: fake_remote
+          endpoints:
+          - lb_endpoints:
+            - endpoint:
+                address:
+                  socket_address: { address: 127.0.0.1, port_value: 10101 }
+    """
+
+class CancelStreamTest {
+
+  init {
+    JniLibrary.loadTestLibrary()
+  }
+
+  private val filterExpectation = CountDownLatch(1)
+  private val runExpectation = CountDownLatch(1)
+
+  class CancelValidationFilter(
+    private val latch: CountDownLatch
+  ) : ResponseFilter {
+    override fun onResponseHeaders(headers: ResponseHeaders, endStream: Boolean): FilterHeadersStatus<ResponseHeaders> {
+      return FilterHeadersStatus.Continue(headers)
+    }
+
+    override fun onResponseData(body: ByteBuffer, endStream: Boolean): FilterDataStatus<ResponseHeaders> {
+      return FilterDataStatus.Continue(body)
+    }
+
+    override fun onResponseTrailers(trailers: ResponseTrailers): FilterTrailersStatus<ResponseHeaders, ResponseTrailers> {
+      return FilterTrailersStatus.Continue(trailers)
+    }
+
+    override fun onError(error: EnvoyError) {}
+
+    override fun onCancel() {
+      latch.countDown()
+    }
+  }
+
+  @Test
+  fun `cancel stream calls onCancel callback`() {
+    val engine = EngineBuilder(Custom(config))
+      .addPlatformFilter(
+        name = "cancel_validation_filter",
+        factory = { CancelValidationFilter(filterExpectation) }
+      )
+      .setOnEngineRunning {}
```

Comment:
this is strange. I will look into this. I'd rather us not have this mystery.

Reply:
Let me add an issue: https://github.com/envoyproxy/envoy-mobile/issues/1429

Reply:
thanks!